### PR TITLE
add GetStartBaseBlock function to optimize token feeder scheduling

### PR DESCRIPTION
**Pull request by leonz789**
## Description

This PR introduces a new `GetStartBaseBlock` function to optimize token feeder scheduling in the oracle module. The function intelligently calculates the optimal `startBaseBlock` for new token feeders to minimize conflicts with existing feeders' quoting windows.
### Problem
New token feeders were scheduled with `startBaseBlock = currentHeight + startAfterBlocks` without considering existing feeders, leading to overlapping quoting windows and reduced oracle efficiency.
### Solution
- **New Function**: `GetStartBaseBlock` analyzes existing feeders and returns optimal offset
- **Smart Scheduling**: Finds blocks with zero conflicts first, then minimum conflicts
----

Closes #XXX

## Release Notes

* **New Features**
  * Token feeders now use dynamic scheduling for new feeder registration, automatically calculating optimal starting positions based on existing feeder configurations.

* **Tests**
  * Added comprehensive test coverage for token feeder scheduling algorithms, including edge cases, overlap detection, conflict scenarios, and multi-feeder configurations.

### CODE REVIEW

File: x/oracle/keeper/params.go [hunk 1]
```diff
@@ -150,3 +156,58 @@ func (k Keeper) RegisterNewTokenAndSetTokenFeeder(ctx sdk.Context, oInfo *types.
 	k.SetParams(ctx, p)
 	return nil
 }
+
+func GetStartBaseBlock(firstQuotingHeight, window, interval uint64, tfs []*types.TokenFeeder) uint64 {
```
Comment (MaxMustermann2) [inline x/oracle/keeper/params.go:200]: Can you add a comment here that describes the overall formula?

File: x/oracle/keeper/params.go [hunk 2]
```diff
@@ -150,3 +156,98 @@ func (k Keeper) RegisterNewTokenAndSetTokenFeeder(ctx sdk.Context, oInfo *types.
...
+		tfStartL := tf.StartBaseBlock
+		// Calculate the first quoting block of this feeder's latest round that's <= firstQuotingHeight
+		if firstQuotingHeight < tfStartL {
+			rounds := (tfStartL - firstQuotingHeight) / tf.Interval
+			// tfStartL = tfStartL - (rounds+1)*tf.Interval
+			if (rounds+1)*tf.Interval > tfStartL {
+				tfStartL = 0
+			} else {
+				tfStartL -= (rounds + 1) * tf.Interval
+			}
```
Comment (MaxMustermann2) [inline x/oracle/keeper/params.go:218]: In my opinion, setting `tfStartL = 0` is not appropriate in the negative case. If `tfStartL` is in the future but less than `tf.Interval` blocks from the `firstQuotingHeight`, setting `tfStartL` to 0 results in the code _believing_ that `tf.StartBaseBlock` is not a "busy" block.


```go
gap = (rounds + 1) * tf.Interval
if gap <= tfStartL {
    tfStartL -= gap
}
```

File: x/oracle/keeper/params.go [hunk 3]
```diff
@@ -150,3 +156,98 @@ func (k Keeper) RegisterNewTokenAndSetTokenFeeder(ctx sdk.Context, oInfo *types.
...
+//
+//	If firstQuotingHeight=5, window=3, interval=30, and offset=1 is returned:
+//	- startBaseBlock = 5 + 1 - 1 = 5
+//	- First quoting block = 5 + 1 = 6
+//	- Quoting window = {6, 7, 8}
+func GetStartBaseBlock(firstQuotingHeight, window, interval uint64, tfs []*types.TokenFeeder) uint64 {
+	blocks := make([]uint64, interval)
+	for _, tf := range tfs {
+		if tf.EndBlock > 0 && tf.EndBlock < firstQuotingHeight {
+			continue
```
Comment (MaxMustermann2) [inline x/oracle/keeper/params.go:209]:
```suggestion
			// already ended
			continue
```

File: x/oracle/keeper/params.go [hunk 4]
```diff
@@ -150,3 +156,98 @@ func (k Keeper) RegisterNewTokenAndSetTokenFeeder(ctx sdk.Context, oInfo *types.
...
+//     The first quoting block will be: startBaseBlock + 1 = firstQuotingHeight + offset
+//
+// Example:
+//
+//	If firstQuotingHeight=5, window=3, interval=30, and offset=1 is returned:
+//	- startBaseBlock = 5 + 1 - 1 = 5
+//	- First quoting block = 5 + 1 = 6
+//	- Quoting window = {6, 7, 8}
+func GetStartBaseBlock(firstQuotingHeight, window, interval uint64, tfs []*types.TokenFeeder) uint64 {
+	blocks := make([]uint64, interval)
```
Comment (MaxMustermann2) [inline x/oracle/keeper/params.go:204]: This variable is common across all `tfs`.
```suggestion
	blocks := make([]uint64, interval)
	scanEnd := firstQuotingHeight + interval
```

File: x/oracle/keeper/params.go [hunk 5]
```diff
@@ -150,3 +156,98 @@ func (k Keeper) RegisterNewTokenAndSetTokenFeeder(ctx sdk.Context, oInfo *types.
...
+					}
+				}
+				tmp++
+			}
+			tfStartL += tf.Interval
+			if tf.EndBlock > 0 && tfStartL >= tf.EndBlock {
+				break
+			}
+		}
+	}
```
Comment (MaxMustermann2) [inline x/oracle/keeper/params.go]:
```suggestion
		for tfStartL < scanEnd {
		    qStart := tfStartL + 1
		    qEnd := tfStartL + window
		    // efficiency fix
		    if qStart < firstQuotingHeight {
                 qStart = firstQuotingHeight
            }
            if qEnd >= scanEnd {
                qEnd = scanEnd - 1
            }
            // readability fix
            for i := qStart; i <= qEnd; i++ {
                blocks[i-firstQuotingHeight]++
            }
            tfStartL += tf.Interval
            if tf.EndBlock > 0 && tfStartL >= tf.EndBlock {
                break
            }
        }   
```

**Review by MaxMustermann2 — CHANGES_REQUESTED** [review]

File: x/oracle/keeper/params.go (hunk 5 above)
Comment (MaxMustermann2) [reply x/oracle/keeper/params.go]: cc @coderabbitai any thoughts on this suggestion?

Comment (MaxMustermann2) [reply x/oracle/keeper/params.go]: Thanks, I edited the inner loop to say `qStart` and `qEnd`.

`scanEnd` is a new variable i defined outside the outermost loop. It is just renamed from `scanUntil`.

File: x/oracle/keeper/params.go (hunk 2 above)
Comment (MaxMustermann2) [reply x/oracle/keeper/params.go:218]: @coderabbitai please provide your inputs on this suggestion

**Review by MaxMustermann2 — APPROVED** [review]

File: x/oracle/keeper/params.go (hunk 5 above)
Comment (leonz789) [reply x/oracle/keeper/params.go]: updated

**Review by cloud8little — APPROVED** [review]
